Fix kernel panic on snapshot destroy for a partition

e-kov opened the pull request:
**Fixed kernel panic on snapshot destroy for a partition**

The kernel panic was happening in case if there are multiple snapshot
devices are for the multiple partitions of the same disk. The problem
was for the kernels 5.9+.

There is a logic for these kernels to replace `block_device_operations`
structure with the driver's tracing function instead of an original
`submit_bio`. This struct belongs to the disk, and it's shared between
partitions of the disk.

The issue was in the access to the freed memory after ours tracing
struct was freed.

Now we are not allocating a new struct when setting up a snapshot for
the 2nd+ partition of the disk. And this struct is freed just when the
last snapshot for some partition of the disk has been destroyed.

The driver has the same behavior with the `make_request` function in
the bio queue for the kernel versions before 5.9. It's replaced with the
tracking function on the first setup snapshot operation for multiple
partitions of the disk. And an original `make_request` is set back when
the last snapshot device has been destroyed respectively.
So, now this behaviour is consistent for all Linux kernel versions.

Fixes https://github.com/elastio/elastio-snap/issues/155

**Implemented new tests for snaps of partitions of the same disk**

The idea is to add a test on the loopback device with 2 or more partitions.
This test should reproduce the bug #155 like this:
1) create snapshots for both partitions of the disk (loopback device);
2) destroy 2nd snapshot device;
3) perform write to the 2nd partition;
4) wait a bit.

These steps are leading to the kernel panic without the fix.

Also added 2 other tests: simple setup test and a test with writes
to all partitions.

**Fixed tests if they are running on machine with '/' on LVM**

There is a test `test_setup_2_volumes` which uses one synthetic device
and 2nd is a root volume of the host machine. This test was failing due
to the multiple and different device names for LVM devices like
`/dev/mapper/ubuntu--vg-ubuntu--lv` and `/dev/dm-0` which are the same
device.

anelson commented on src/elastio-snap.c:
```diff
@@ -4290,22 +4313,30 @@ static int __tracer_setup_tracing(struct snap_device *dev, unsigned int minor){
-			LOG_ERROR(ret, "error allocating tracing ops");
-			goto error;
-		}
+			LOG_DEBUG("allocating tracing ops for device with minor %i", minor);
+			dev->sd_tracing_ops = kmalloc(sizeof(struct block_device_operations), GFP_KERNEL);
+			if (!dev->sd_tracing_ops) {
+				ret = -ENOMEM;
+				LOG_ERROR(ret, "error allocating tracing ops");
+				goto error;
+			}
 
```
Why does the indentation look off here?  Is it an artifact of the GitHub UI or did you really indent this code more than the surrounding code?

e-kov replied:
Probably it's an artifact of the GitHub UI.  The tabs are used in this source file for indentation.  I like more to use spaces, because tabs may look different in different editors and UIs.  But for the code stile consistency we are still using tabs here.
The max indent is 5 tabs on lines 4326-4328.  And it's hard to tell how many spaces are used for 1 tab here in the GitHub Sometimes it looks like 6 and sometimes like 7 spaces. :man_shrugging:

anelson commented on tests/devicetestcase_multipart.py:
```diff
@@ -0,0 +1,60 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+#
+# Copyright (C) 2020 Elastio Software Inc.
```
Should be 2022 unless this code was really written two years ago.

e-kov replied:
Done

anelson commented on tests/test_multipart.py:
```diff
@@ -0,0 +1,110 @@
+# SPDX-License-Identifier: GPL-2.0-only
+
+#
+# Copyright (C) 2020 Elastio Software Inc.
```
2022

e-kov replied:
Done

anelson reviewed (commented):
Nice, this looks like a fix that was really tough to track down!

Minor comments only.

e-kov commented on src/elastio-snap.c:
```diff
@@ -4231,25 +4236,43 @@ static void minor_range_include(unsigned int minor){
+				break;
+			}
+		}
+		if (free_tracing_ops) kfree(dev->sd_tracing_ops);
```
replaced

e-kov commented:
@anelson @vsazhenyuk-softheme since your last review:
- Addressed @skypodolsky's [comment](https://github.com/elastio/elastio-snap/pull/160#discussion_r948051806) - used refcounts for the tracing block device_operation_struct.
- Fixed mistake in `elio-test.sh` when using external devices for the LVM/raid scenarios. And a few following issues are fixed too.
- Increased timeouts for the LVM/raid scenarios and decreased qcow2 disk size from 1G to 256M for `arm64` due to several timeout errors. (the storage in ec2 is slow comparatively with the physical servers).

e-kov commented on .github/workflows/ci.yml:
```diff
@@ -205,9 +205,10 @@ jobs:
           [ ${ARCH} != "x86_64" ] && VIRSH_FLAGS="--config" || true
           for i in ${!TEST_IMAGES[*]}; do
-            qemu-img create -f qcow2 ${TEST_IMAGES[i]} 1G
+            qemu-img create -f qcow2 ${TEST_IMAGES[i]} $SZ
```
[It's 1G for `86_64` and 256M for `aarch64`.
https://github.com/elastio/elastio-snap/blob/fix/155-multipartitions/.github/workflows/ci.yml#L208

skypodolsky reviewed (approved):
Looks good to me. Good work :+1:
